src/string.rs

```diff
@@ -60,17 +58,32 @@ impl<'gc> Debug for String<'gc> {
 
 impl<'gc> String<'gc> {
     pub fn new(mc: MutationContext<'gc, '_>, s: &[u8]) -> String<'gc> {
+        String::new_mapped(mc, s, |from, idx| from[idx])
+    }
+
+    pub fn new_mapped<F>(mc: MutationContext<'gc, '_>, s: &[u8], f: F) -> String<'gc>
```

Comment:
This is pretty specific and a bit over-kill for the current optimization level.  I'm aware that String construction is not the fastest, but what's really blocking it is dynamically sized types backed by a Gc pointer.  This is waiting on some details of the allocator system to be made stable.

An alternative would be to be able to construct a String from a `Box<[u8]>` for now, so if you need to allocate a `Vec<u8>` as an intermediary then it can just directly become that type of string.

Another alternative, maybe better, is to implement `FromIterator<u8>`.